WordPress/gutenberg
# Legacy widget rendering endpoint

## Description
Follows up on: https://github.com/WordPress/wordpress-develop/pull/1508

This adds a `/wp/v2/widget-types/<widget type>/render` endpoint which accepts POST data and returns the same output as the previous iframe-oriented GET endpoint.

This PR is for the Gutenberg plugin and not for core so that we the endpoint may be safely used in other Gutenberg PRs.

The polyfill approach draws on learnings from https://github.com/WordPress/gutenberg/issues/33810 and may not be perfect – let's use this PR as a testing ground. One shortcoming is that I can't think of a good way to apply the following feedback from @noisysocks:

> Did you consider incorporating this functionality into the existing /widget-types/:id/encode endpoint? /encode is an existing RPC-style endpoint that already accepts instance as an argument and returns preview in the response. The only difference is that /encode does not output all of the wrapping HTML including scripts and styles and whatnot. Perhaps it could, though. I'm not sure that we're using the preview from /encode anywhere. If it's possible to have fewer endpoints (especially RPC-style ones) then I'd prefer that! 😀

I like the idea of re-using the `/encode` endpoint, but I have no idea how to polyfill that. I'd rather avoid overriding the entire endpoint with an updated copy living in this repo – it could send us to a land of two-way synchronization between `wordpress-develop` and `gutenberg` which I think is much more complex than one-way backporting workflow.

## How has this been tested?

1. Go to `/wp-admin/widgets.php`
2. Paste the following script into your devtools:

```js
wp.apiFetch({
      "path": "/index.php?rest_route=%2Fwp%2Fv2%2Fwidget-types%2Ftext%2Frender&context=edit&per_page=100&_locale=user",
      "method": "POST",
      "data": {"id_base":"text","instance":{"encoded":"YTo0OntzOjU6InRpdGxlIjtzOjU6InRlc3QyIjtzOjQ6InRleHQiO3M6NToidGVzdDMiO3M6NjoiZmlsdGVyIjtiOjE7czo2OiJ2aXN1YWwiO2I6MTt9","hash":"6bd53d501ad76cfbc50ef417d78dc9f1","raw":{"title":"test2","text":"test3","filter":true,"visual":true}}}
})
.then(console.log.bind(console))
.catch(console.error.bind(console));
```

3. Confirm the response was successful and the widget got rendered as expected.

## Alternatives considered
* `ServerSideRender` – we need to render a full HTML document with `wp_head()`, CSS links, maybe some additional scripts, and this is much more that I would like to return from a block rendering API.
* Updating the existing `admin_init` hook handler to work with `POST` requests – it's tricky and doesn't seem right, I don't think we should stretch hooks like that. This clearly is an API request so let's treat it as such.
